### ISSUE TITLE
Pyright ignores parameter type annotation and assigns a different type in method body.

### ISSUE THREAD

**Issue by kputnam** [microsoft/pyright#3694]
**Describe the bug**

In the following example, pyright assigns a type different from the annotation to a parameter. This seems to happen because the parameter annotation uses a type variable that's scoped to the method's (generic) class.

**To Reproduce**

```python
from typing import Callable, Generic, TypeVar

T = TypeVar('T')
U = TypeVar('U')
V = TypeVar('V')

class F(Generic[T, U]):
    f: Callable[[T], U]

    def __init__(self, f: Callable[[T], U]):
        self.f = f

    def __call__(self, x: T) -> U:
        return self.f(x)

    def __add__(self, other: 'F[U, V]') -> 'F[T, V]':
        f: Callable[[T], U] = self.f
        g: Callable[[U], V] = other.f
        return F(lambda x: g(f(x)))
```

**Expected behavior**

The statement at `g: ...` should type check.

**Screenshots or Code**

![Screen Shot 2022-07-13 at 9 36 58 PM](https://user-images.githubusercontent.com/235798/178886128-033ebe27-61b0-40dc-8460-a0cab2d04f77.png)

**VS Code extension or command-line**

This happens in VS Code, neovim (using williamboman/nvim-lsp-installer and neovim/nvim-lspconfig)
, and also on the command line (npm install -g pyright).

```
» pyright --version
pyright 1.1.237

» pyright
No configuration file found.
No pyproject.toml file found.
stubPath /Users/kputnam/test/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
/Users/kputnam/test/test.py
  /Users/kputnam/test/test.py:19:31 - error: Expression of type "(V@__add__) -> V@__add__" cannot be assigned to declared type "(U@F) -> V@__add__"
    Type "(V@__add__) -> V@__add__" cannot be assigned to type "(U@F) -> V@__add__"
      Parameter 1: type "U@F" cannot be assigned to type "V@__add__"
        Type "U@F" cannot be assigned to type "V@__add__" (reportGeneralTypeIssues)
  /Users/kputnam/test/test.py:20:32 - error: Argument of type "T@F" cannot be assigned to parameter of type "U@F"
    Type "T@F" cannot be assigned to type "U@F" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
Completed in 0.633sec
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.261, which I just published. It will also be included in an upcoming release of pylance.
